ssl/statem/extensions.c

```diff
@@ -890,40 +1124,77 @@ int tls_construct_extensions(SSL_CONNECTION *s, WPACKET *pkt,
         return 0;
     }
 
-    for (i = 0, thisexd = ext_defs; i < OSSL_NELEM(ext_defs); i++, thisexd++) {
-        EXT_RETURN (*construct)(SSL_CONNECTION *s, WPACKET *pkt,
-                                unsigned int context,
-                                X509 *x, size_t chainidx);
-        EXT_RETURN ret;
+#ifndef OPENSSL_NO_ECH
+    /*
+     * Two passes - we first construct the to-be-ECH-compressed
+     * extensions, and then go around again constructing those that
+     * aren't to be ECH-compressed. We need to ensure this ordering
+     * so that all the ECH-compressed extensions are contiguous
+     * in the encoding. The actual compression happens later in
+     * ech_encode_inner().
+     */
+    for (pass = 0; pass <= 1; pass++)
```

Comment:
So we're going to do two passes even if ECH has not been configured? Also seems to be the case we're doing two passes even if we're not in the ClientHello context?

Reply:
We do two passes yes, but the same amount of processing - we just do the extensions in a different order (the ones to be ECH compressed in 1st pass, then the others) in the 2nd pass. The two pass thing is really just a way to ensure the to-be ECH-compresses exts are contiguous without putting onerous constraints on other code.